src/Utilities/ResourceCollection.h

```diff
@@ -76,5 +77,41 @@ class ResourceCollectionLock
   const bool active;
 };
 
+
```

Comment:
Some documentation of why ResourceCollectionTeamLock is useful is needed.

At first glance it seems 
```c++
{
ResourceCollectionLock(res_coll, batched_things.get_leader())
callNeedingResource(batched_things);
}
```
Would do that same thing.

But it seems like what this really does is give the leader's acquire_resource call access to the whole refvector and not just the leader. 

I can't judge yet whether adjusting/checking the resource state just in time in a particular call or a mechanism like this is better, I'm guessing you added this because at the moment both are needed to refactor existing code. Perhaps we will want to prefer one for most resource consumers.

Reply:
The current code uses ResourceCollectionTeamLock for ParticleSet for my exploration but the reset has not been changed yet. It is a bit restrictive and needs acuqire/release APIs to be conformal. But maybe this restrictive way is better to discipline the implementation more.